File: apps/api/app/core/pipeline_stage_colors.py
"""Pipeline stage color helpers."""

from __future__ import annotations

import re

DEFAULT_CUSTOM_STAGE_COLOR = "#6b7280"
FALLBACK_GRAY_STAGE_COLORS = {DEFAULT_CUSTOM_STAGE_COLOR}

CUSTOM_STAGE_COLOR_PRESETS: dict[str, list[str]] = {
    "intake": ["#2563eb", "#0ea5e9", "#14b8a6", "#22c55e", "#8b5cf6", "#f59e0b"],
    "post_approval": ["#0f766e", "#0891b2", "#4f46e5", "#8b5cf6", "#db2777", "#ea580c"],
    "paused": ["#b4536a"],
    "terminal": ["#ef4444", "#dc2626"],
}

KEYWORD_STAGE_COLOR_RULES: tuple[tuple[tuple[str, ...], str], ...] = (
    (("docusign", "docu", "signature", "consent"), "#f59e0b"),
    (("insurance",), "#0891b2"),
    (("medical", "medically", "clinic"), "#14b8a6"),
    (("legal", "contract"), "#6366f1"),
    (("transfer", "cycle"), "#0d9488"),
    (("pbo",), "#db2777"),
    (("anatomy", "heartbeat", "hcg", "pregnancy", "pregnant", "ob"), "#16a34a"),
    (("interview", "consult", "screen"), "#a855f7"),
    (("application", "packet", "submit", "submitted"), "#8b5cf6"),
    (("review",), "#f59e0b"),
    (("qualif",), "#10b981"),
    (("contact", "outreach"), "#0ea5e9"),
)


def _normalize_hex_color(color: str | None) -> str | None:
    if not color:
        return None
    normalized = color.strip()
    if re.fullmatch(r"#[0-9a-fA-F]{6}", normalized):
        return normalized
    return None


def _normalize_stage_text(*parts: str | None) -> str:
    joined = " ".join(part for part in parts if part)
    return re.sub(r"[^a-z0-9]+", " ", joined.lower()).strip()


def _fallback_palette_index(stage_type: str | None, stage_key: str | None, order: int | None) -> int:
    palette = CUSTOM_STAGE_COLOR_PRESETS.get(stage_type or "", [DEFAULT_CUSTOM_STAGE_COLOR])
    if order and order > 0:
        return (order - 1) % len(palette)
    seed = _normalize_stage_text(stage_key, stage_type)
    return sum((index + 1) * ord(char) for index, char in enumerate(seed)) % len(palette)


def suggest_stage_color(
    *,
    label: str | None,
    slug: str | None,
    stage_key: str | None,
    stage_type: str | None,
    order: int | None = None,
) -> str:
    normalized_text = _normalize_stage_text(label, slug, stage_key)
    for keywords, color in KEYWORD_STAGE_COLOR_RULES:
        if any(keyword in normalized_text for keyword in keywords):
            return color

    palette = CUSTOM_STAGE_COLOR_PRESETS.get(stage_type or "", [DEFAULT_CUSTOM_STAGE_COLOR])
    return palette[_fallback_palette_index(stage_type, stage_key or slug, order)]


def resolve_stage_color(
    *,
    color: str | None,
    label: str | None,
    slug: str | None,
    stage_key: str | None,
    stage_type: str | None,
    order: int | None = None,
    is_locked: bool = False,
) -> str:
    normalized_color = _normalize_hex_color(color)
    if normalized_color and normalized_color.lower() not in FALLBACK_GRAY_STAGE_COLORS:
        return normalized_color
    if is_locked:
        return normalized_color or DEFAULT_CUSTOM_STAGE_COLOR
    return suggest_stage_color(
        label=label,
        slug=slug,
        stage_key=stage_key,
        stage_type=stage_type,
        order=order,
    )

File: apps/web/lib/pipeline-stage-colors.ts
import type { StageType } from "@/lib/api/pipelines"

const DEFAULT_CUSTOM_STAGE_COLOR = "#6b7280"
const FALLBACK_GRAY_STAGE_COLORS = new Set([DEFAULT_CUSTOM_STAGE_COLOR])

export const CUSTOM_STAGE_COLOR_PRESETS: Record<StageType, string[]> = {
    intake: ["#2563eb", "#0ea5e9", "#14b8a6", "#22c55e", "#8b5cf6", "#f59e0b"],
    post_approval: ["#0f766e", "#0891b2", "#4f46e5", "#8b5cf6", "#db2777", "#ea580c"],
    paused: ["#b4536a"],
    terminal: ["#ef4444", "#dc2626"],
}

const KEYWORD_STAGE_COLOR_RULES: Array<{ keywords: string[]; color: string }> = [
    { keywords: ["docusign", "docu", "signature", "consent"], color: "#f59e0b" },
    { keywords: ["insurance"], color: "#0891b2" },
    { keywords: ["medical", "medically", "clinic"], color: "#14b8a6" },
    { keywords: ["legal", "contract"], color: "#6366f1" },
    { keywords: ["transfer", "cycle"], color: "#0d9488" },
    { keywords: ["pbo"], color: "#db2777" },
    { keywords: ["anatomy", "heartbeat", "hcg", "pregnancy", "pregnant", "ob"], color: "#16a34a" },
    { keywords: ["interview", "consult", "screen"], color: "#a855f7" },
    { keywords: ["application", "packet", "submit", "submitted"], color: "#8b5cf6" },
    { keywords: ["review"], color: "#f59e0b" },
    { keywords: ["qualif"], color: "#10b981" },
    { keywords: ["contact", "outreach"], color: "#0ea5e9" },
]

type StageColorInput = {
    color?: string | null
    label?: string | null
    slug?: string | null
    stage_key?: string | null
    stage_type?: StageType | string | null
    order?: number | null
    is_locked?: boolean | null
}

function normalizeHexColor(color: string | null | undefined) {
    const normalized = color?.trim()
    return normalized && /^#[0-9a-fA-F]{6}$/.test(normalized) ? normalized : null
}

function normalizeStageText(...parts: Array<string | null | undefined>) {
    return parts
        .filter(Boolean)
        .join(" ")
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, " ")
        .trim()
}

function getFallbackPaletteColor(stageType: StageType | string | null | undefined, stageKey: string | null | undefined, order?: number | null) {
    const palette = CUSTOM_STAGE_COLOR_PRESETS[(stageType as StageType) ?? "intake"] ?? [DEFAULT_CUSTOM_STAGE_COLOR]
    if (order && order > 0) {
        return palette[(order - 1) % palette.length] ?? DEFAULT_CUSTOM_STAGE_COLOR
    }
    const seed = normalizeStageText(stageKey, stageType)
    const hash = Array.from(seed).reduce((total, char, index) => total + (index + 1) * char.charCodeAt(0), 0)
    return palette[hash % palette.length] ?? DEFAULT_CUSTOM_STAGE_COLOR
}

export function suggestStageColor(stage: StageColorInput) {
    const normalizedText = normalizeStageText(stage.label, stage.slug, stage.stage_key)
    for (const rule of KEYWORD_STAGE_COLOR_RULES) {
        if (rule.keywords.some((keyword) => normalizedText.includes(keyword))) {
            return rule.color
        }
    }
    return getFallbackPaletteColor(stage.stage_type, stage.stage_key ?? stage.slug, stage.order)
}

export function resolveStageColor(stage: StageColorInput) {
    const normalizedColor = normalizeHexColor(stage.color)
    if (normalizedColor && !FALLBACK_GRAY_STAGE_COLORS.has(normalizedColor.toLowerCase())) {
        return normalizedColor
    }
    if (stage.is_locked) {
        return normalizedColor ?? DEFAULT_CUSTOM_STAGE_COLOR
    }
    return suggestStageColor(stage)
}

export function shouldAutoRefreshStageColor(stage: StageColorInput) {
    if (stage.is_locked) return false
    const normalizedColor = normalizeHexColor(stage.color)
    const autoKey = (stage.stage_key ?? stage.slug ?? "").toLowerCase()
    return (
        !normalizedColor
        || FALLBACK_GRAY_STAGE_COLORS.has(normalizedColor.toLowerCase())
        || stage.label === "New Stage"
        || autoKey.startsWith("custom_stage")
        || normalizedColor === suggestStageColor(stage)
    )
}
